src/local_executor.rs

```diff
@@ -0,0 +1,631 @@
+        // variables in the future.
+        let (runnable, task) = unsafe {
+            Builder::new()
+                .propagate_panic(true)
+                .spawn_unchecked(|()| future, self.schedule())
+        };
+        entry.insert(runnable.waker());
+
+        runnable.schedule();
+        task
+    }
+
+    /// Attempts to run a task if at least one is scheduled.
+    ///
+    /// Running a scheduled task means simply polling its future once.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use async_executor::LocalExecutor;
+    ///
+    /// let ex = LocalExecutor::new();
+    /// assert!(!ex.try_tick()); // no tasks to run
+    ///
+    /// let task = ex.spawn(async {
+    ///     println!("Hello world");
+    /// });
+    /// assert!(ex.try_tick()); // a task was found
+    /// ```
+    pub fn try_tick(&self) -> bool {
+        self.state().try_tick()
+    }
+
+    /// Runs a single task.
+    ///
+    /// Running a task means simply polling its future once.
+    ///
+    /// If no tasks are scheduled when this method is called, it will wait until one is scheduled.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use async_executor::LocalExecutor;
+    /// use futures_lite::future;
+    ///
+    /// let ex = LocalExecutor::new();
+    ///
+    /// let task = ex.spawn(async {
+    ///     println!("Hello world");
+    /// });
+    /// future::block_on(ex.tick()); // runs the task
+    /// ```
+    pub async fn tick(&self) {
+        self.state().tick().await;
+    }
+
+    /// Runs the executor until the given future completes.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use async_executor::LocalExecutor;
+    /// use futures_lite::future;
+    ///
+    /// let local_ex = LocalExecutor::new();
+    ///
+    /// let task = local_ex.spawn(async { 1 + 2 });
+    /// let res = future::block_on(local_ex.run(async { task.await * 2 }));
+    ///
+    /// assert_eq!(res, 6);
+    /// ```
+    pub async fn run<T>(&self, future: impl Future<Output = T>) -> T {
+        self.state().run(future).await
+    }
+
+    /// Returns a function that schedules a runnable task when it gets woken up.
+    fn schedule(&self) -> impl Fn(Runnable) + 'static {
+        let state = self.state_as_rc();
+
+        move |runnable| {
+            {
+                // SAFETY: All UnsafeCell accesses to queue are tightly scoped, and because
+                // `LocalExecutor` is !Send, there is no way to have concurrent access to the
+                // values in `State`, including the queue field.
+                let queue = unsafe { &mut *state.queue.get() };
+                queue.push_front(runnable);
+            }
+            state.notify();
+        }
+    }
+
+    /// Returns a pointer to the inner state.
+    #[inline]
+    pub(crate) fn state_ptr(&self) -> *const State {
+        #[cold]
+        fn alloc_state(cell: &Cell<*mut State>) -> *mut State {
```

Comment:
Why do we delay the allocation of the state? How beneficial is this from a performance perspective (given that it trades a potential "away optimization" of state allocation in case the `LocalExecutor` gets destroyed without usage for a branch on every state access (minus a few where the compiler can perhaps infer that the state is non-null / initialized))

Reply:
It allows `LocalExecutor::new` to be const, without it this would be a breaking change and would require a major version bump.

Reply:
As far as I'm aware, this sounds correct.

Reply:
Okay, for the next major version bump it would be nice to pursue a version which doesn't use the internal `Cell` wrapping, at least for the `LocalExecutor` version, where "difficulty of sharing across threads" isn't that much of a concern.

Reply:
Went ahead and implemented the pin change, and filed #146 for the non-local version.